xbmc/pvr/providers/PVRProvider.h

```diff
@@ -0,0 +1,232 @@
+/*
+ *  Copyright (C) 2012-2018 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#pragma once
+
+#include "addons/kodi-dev-kit/include/kodi/c-api/addon-instance/pvr/pvr_providers.h"
+#include "threads/CriticalSection.h"
+#include "utils/ISerializable.h"
+
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace PVR
+{
+
+enum class ProviderUpdateMode
+{
+  BY_CLIENT,
+  BY_DATABASE
+};
+
+static constexpr int PVR_PROVIDER_ADDON_UID = -1;
+static constexpr int PVR_PROVIDER_INVALID_DB_ID = -1;
+
+class CPVRProvider final : public ISerializable
+{
+public:
+  CPVRProvider(int iUniqueId, int iClientId);
+  CPVRProvider(const PVR_PROVIDER& provider, int iClientId);
+  CPVRProvider(int iClientId,
+               const std::string& addonProviderName,
+               const std::string& addonIconPath,
+               const std::string& addonThumbPath);
+
+  bool operator==(const CPVRProvider& right) const;
+  bool operator!=(const CPVRProvider& right) const;
+
+  void Serialize(CVariant& value) const override;
+
+  /*!
+   * @brief The database id of this provider
+   *
+   * A unique identifier for this provider.
+   * It can be used to find the same provider on this clients backend
+   *
+   * @return The database id of this provider
+   */
+  int GetDatabaseId() const;
+
+  /*!
+   * @brief Set the database id of this provider
+   * @param iDatabaseId The new ID.
+   * @return True if the something changed, false otherwise.
+   */
+  bool SetDatabaseId(int iDatabaseId);
+
+  /*!
+   * @brief A unique identifier for this provider.
+   *
+   * A unique identifier for this provider.
+   * It can be used to find the same provider on this clients backend
+   *
+   * @return The Unique ID.
+   */
+  int GetUniqueId() const;
+
+  /*!
+   * @return The identifier of the client that supplies this provider.
+   */
+  int GetClientId() const;
+
+  /*!
+   * @return The name of the provider. Can be user provided or the backend name
+   */
+  std::string GetProviderName() const;
+
+  /*!
+   * @brief Set the name of the provider.
+   * @param providerName The new name of the provider.
+   * @return True if the something changed, false otherwise.
+   */
+  bool SetProviderName(const std::string& iProviderName);
+
+  /*!
+   * @brief Checks whether this provider has a known type
+   * @return True if this provider has a type other than unknown, false otherwise
+   */
+  bool HasProviderType() const { return m_providerType != PVR_PROVIDER_TYPE_UNKNOWN; }
+
+  /*!
+   * @brief Gets the type of this provider.
+   * @return the type of this provider.
+   */
+  PVR_PROVIDER_TYPE GetProviderType() const;
+
+  /*!
+   * @brief Sets the type of this provider.
+   * @param providerType the new provider type.
+   * @return True if the something changed, false otherwise.
+   */
+  bool SetProviderType(PVR_PROVIDER_TYPE providerType);
```

Comment:
Remove `Provider`

Using the bare addon API C struct here breaks encapsulation. This is not a clean design.

Reply:
This is an enum not a struct. In that case should we be using defines and enums from the C API in other parts of kodi?

Reply:
> This is an enum not a struct.

sure

> In that case should we be using defines and enums from the C API in other parts of kodi?

 Ideally not. But as you said, strict encapsulation would require a lot of copied types. Possibly not worth the effort.

Reply:
Ya, I'll leave it.